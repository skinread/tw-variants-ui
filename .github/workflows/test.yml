# This workflow will do a clean installation of node dependencies, cache/restore them, build the source code and run tests across different versions of node
# For more information see: https://docs.github.com/en/actions/automating-builds-and-tests/building-and-testing-nodejs

name: CI Tests

on:
  pull_request:
    branches: ["main","develop"]

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        node-version: [20.x]
    steps:
    - name: Checkout
      uses: actions/checkout@v4
    - name: Setup Node ${{ matrix.node-version }}
      uses: actions/setup-node@v4
      with:
        node-version: ${{ matrix.node-version }}
        cache: 'npm'
        cache-dependency-path: "**/package-lock.json"
    - name: Cache or restore dependencies
      uses: actions/cache@v4
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
        restore-keys: ${{ runner.os }}-npm-
    - name: Install dependencies
      run: npm ci

    - name: Check formatting and linting
      run: npm run prettier && npm run lint

    - name: Check package build
      run: npm run build

    - name: Get Playwright version
      run: echo "PLAYWRIGHT_VERSION=$(jq -r '.devDependencies["@playwright/test"]' package.json)" >> $GITHUB_ENV
    - name: Cache or restore Playwright binaries
      uses: actions/cache@v4
      with:
        path: |
          ~/.cache/ms-playwright
        key: ${{ runner.os }}-playwright-${{ env.PLAYWRIGHT_VERSION }}
        restore-keys: ${{ runner.os }}-playwright-
    - name: Install Playwright
      if: steps.playwright-cache.outputs.cache-hit != 'true'
      run: npm run pretest
    - name: Storybook test build
      run: npm run storybook:build -- --test
    - name: Storybook test suite
      run: npm run test:ci
