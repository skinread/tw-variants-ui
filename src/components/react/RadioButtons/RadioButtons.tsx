import { forwardRef } from 'react';
import { inputVariants, validityFromProps } from '@common';
import { type CommonFieldProps } from '@components/react';

type RadioButtonsPropsFiltered = Omit<
  React.ComponentPropsWithoutRef<'input'>,
  'name' | 'value' | 'defaultChecked'
>;

type CommonFieldPropsFiltered = Pick<CommonFieldProps, 'name' | 'hasError' | 'id'>;

export interface RadioButtonsProps extends CommonFieldPropsFiltered, RadioButtonsPropsFiltered {
  /**
   * Array of labels and values `{ label: "Option 1", value: "1"}`
   */
  options: {
    label: string;
    value: string;
  }[];
  /**
   * Indicate by the value which item in the options array to preselect
   */
  defaultChecked?: string;
}

/**
 * (Work in progress) Styled radio button group expecting 2 or more options.
 * React `ref` enabled. Also exposes intrinsic `<input>` attributes.
 */
export const RadioButtons = forwardRef<HTMLInputElement, RadioButtonsProps>(
  function VuiRadioButtons(props, ref) {
    const { defaultChecked, hasError, options, ...elemAttrs } = props;
    const {
      wrapper: wrapperStyle,
      label: labelStyle,
      input: inputStyle,
      labelText: labelTextStyle,
    } = inputVariants({ type: 'radio', validity: validityFromProps({ hasError }) });

    return options.map((radio, idx) => (
      <div className={wrapperStyle()} key={idx}>
        <label className={labelStyle()}>
          <input
            type="radio"
            value={radio.value}
            defaultChecked={defaultChecked === radio.value ? true : undefined}
            className={inputStyle()}
            {...elemAttrs}
            ref={ref}
          />
          <span className={labelTextStyle()}>{radio.label}</span>
        </label>
      </div>
    ));
  }
);
