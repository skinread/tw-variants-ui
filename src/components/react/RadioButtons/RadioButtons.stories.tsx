import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, within } from '@storybook/test';

import { RadioButtons } from './RadioButtons';

const meta = {
  title: 'React/Radio Buttons',
  component: RadioButtons,
  tags: ['autodocs'],
  args: {
    options: [
      {
        label: 'Option one',
        value: '1',
      },
      {
        label: 'Option two',
        value: '2',
      },
      {
        label: 'Option three',
        value: '3',
      },
    ],
    name: 'demo-radio-buttons',
    hasError: undefined,
  },
  argTypes: {
    onChange: {
      action: 'changed',
    },
  },
} satisfies Meta<typeof RadioButtons>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {
  args: {
    defaultChecked: '1',
  },
};

export const InteractionTest: Story = {
  play: async ({ args, canvasElement, step }) => {
    const canvas = within(canvasElement);
    const inputs = canvas.getAllByRole('radio');

    await step('Initial state', async () => {
      await expect(inputs).toHaveLength(args.options.length);

      for (const [index, input] of inputs.entries()) {
        await expect(input.parentElement).toHaveTextContent(`${args.options[index]?.label}`);
        await expect(input).toHaveAttribute('name', args.name);
        await expect(input).toHaveAttribute('value', args.options[index]?.value);
        await expect(input).not.toBeChecked();
      }
    });

    await step('Checked state', async () => {
      for (const input of inputs) {
        await userEvent.click(input);
        await expect(args.onChange).toHaveBeenCalled();
        await expect(input).toBeChecked();
      }
    });

    await step('Keyboard interaction', async () => {
      await userEvent.keyboard('[ArrowLeft]');
      await expect(args.onChange).toHaveBeenCalled();
      await expect(inputs[0]).toBeChecked();
      await userEvent.keyboard('[ArrowLeft]');
      await expect(args.onChange).toHaveBeenCalled();
      await expect(inputs[1]).toBeChecked();
    });
  },
};
