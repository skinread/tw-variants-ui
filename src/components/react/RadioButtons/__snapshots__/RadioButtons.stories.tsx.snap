// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Radio Buttons Basic smoke-test 1`] = `
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="1"
           checked
    >
    <span class="label-text text-base ml-2.5">
      Option one
    </span>
  </label>
</div>
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="2"
    >
    <span class="label-text text-base ml-2.5">
      Option two
    </span>
  </label>
</div>
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="3"
    >
    <span class="label-text text-base ml-2.5">
      Option three
    </span>
  </label>
</div>
`;

exports[`React/Radio Buttons InteractionTest play-test 1`] = `
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="1"
    >
    <span class="label-text text-base ml-2.5">
      Option one
    </span>
  </label>
</div>
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="2"
    >
    <span class="label-text text-base ml-2.5">
      Option two
    </span>
  </label>
</div>
<div class="form-control flex-row">
  <label class="label items-start px-0 cursor-pointer">
    <input type="radio"
           class="placeholder:text-subtle radio radio-primary w-[--vui-radio-size] h-[--vui-radio-size] bg-white"
           name="demo-radio-buttons"
           value="3"
    >
    <span class="label-text text-base ml-2.5">
      Option three
    </span>
  </label>
</div>
`;
